src/qt_gui/keyboardcontrolswindow.cpp

```diff
@@ -0,0 +1,620 @@
+    ui->DPadUpkeySequenceEdit->setKeySequence(
+        convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::DPadUp_Key]));
+    ui->DPadDownkeySequenceEdit->setKeySequence(
+        convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::DPadDown_Key]));
+    ui->L2keySequenceEdit->setKeySequence(convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::L2_Key]));
+    ui->L1keySequenceEdit->setKeySequence(convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::L1_Key]));
+    ui->CrossKeySequenceEdit->setKeySequence(
+        convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::Cross_Key]));
+    ui->R2KeySequenceEdit->setKeySequence(convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::R2_Key]));
+    ui->CircleKeySequenceEdit->setKeySequence(
+        convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::Circle_Key]));
+    ui->R1KeySequenceEdit->setKeySequence(convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::R1_Key]));
+    ui->SquareKeySequenceEdit->setKeySequence(
+        convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::Square_Key]));
+    ui->TriangleKeySequenceEdit->setKeySequence(
+        convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::Triangle_Key]));
+
+    QObject::connect(ui->applyButton, &QPushButton::clicked,
+                     [this]() { validateAndSaveKeyBindings(); });
+
+    QObject::connect(ui->cancelButton, &QPushButton::clicked, [this]() { this->close(); });
+
+    QList<QKeySequenceEdit*> keySequenceEdits = {
+        ui->StartKeySequenceEdit,        ui->SelectKeySequenceEdit,
+        ui->LAnalogDownkeySequenceEdit,  ui->LAnalogLeftkeySequenceEdit,
+        ui->LAnalogUpkeySequenceEdit,    ui->LAnalogRightkeySequenceEdit,
+        ui->PSkeySequenceEdit,           ui->RAnalogDownkeySequenceEdit,
+        ui->RAnalogLeftkeySequenceEdit,  ui->RAnalogUpkeySequenceEdit,
+        ui->RAnalogRightkeySequenceEdit, ui->DPadLeftkeySequenceEdit,
+        ui->DPadRightkeySequenceEdit,    ui->DPadUpkeySequenceEdit,
+        ui->DPadDownkeySequenceEdit,     ui->L2keySequenceEdit,
+        ui->L1keySequenceEdit,           ui->CrossKeySequenceEdit,
+        ui->R2KeySequenceEdit,           ui->CircleKeySequenceEdit,
+        ui->R1KeySequenceEdit,           ui->SquareKeySequenceEdit,
+        ui->TriangleKeySequenceEdit};
+
+    for (auto edit : keySequenceEdits) {
+        edit->setStyleSheet("QLineEdit { qproperty-alignment: AlignCenter; }");
+    }
+}
+
+KeyboardControlsWindow::~KeyboardControlsWindow() {
+    delete ui;
+}
+
+std::map<Uint32, KeysMapping> KeyboardControlsWindow::getKeysMapping() {
+    return m_keysMap;
+}
+
+void KeyboardControlsWindow::validateAndSaveKeyBindings() {
+    int nOfUnconfiguredButtons = 0;
+    for (auto& keyEdit : m_listOfKeySequenceEdits) {
+        auto keySequence = keyEdit->keySequence();
+
+        if (keySequence == QKeySequence()) {
+            keyEdit->setStyleSheet("background-color: red; qproperty-alignment: AlignCenter;");
+            QTimer::singleShot(2000, keyEdit, [keyEdit]() {
```

Comment:
2000 is a magic number, make a constexpr constant for it

Reply:
you're right!